workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs

```diff
@@ -0,0 +1,236 @@
+{
+    public static class ComponentDiffStorageGenerator
+    {
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+        public static string Generate(UnityComponentDetails componentDetails, string qualifiedNamespace)
+        {
+            var eventDetailsList = componentDetails.EventDetails;
+
+            return CodeWriter.Populate(cgw =>
+            {
+                cgw.UsingDirectives(
+                    "System.Collections.Generic",
+                    "System",
+                    "Improbable.Gdk.Core",
+                    "Improbable.Worker.CInterop"
+                );
+
+                cgw.Namespace(qualifiedNamespace, ns =>
+                {
+                    ns.Type($"public partial class {componentDetails.ComponentName}", partial =>
+                    {
+                        Logger.Trace($"Generating {qualifiedNamespace}.{componentDetails.ComponentName}.DiffComponentStorage class.");
+
+                        var classDefinition = new StringBuilder("public class DiffComponentStorage : IDiffUpdateStorage<Update>, IDiffComponentAddedStorage<Update>, IDiffAuthorityStorage");
+
+                        foreach (var ev in eventDetailsList)
+                        {
+                            var eventType = $"{ev.EventName}.Event";
+                            classDefinition.Append($"{Environment.NewLine}    , IDiffEventStorage<{eventType}>");
+                        }
+
+                        partial.Type(classDefinition.ToString(), storage =>
+                        {
+                            storage.Line(@"
+private readonly HashSet<EntityId> entitiesUpdated = new HashSet<EntityId>();
+
+private List<EntityId> componentsAdded = new List<EntityId>();
+private List<EntityId> componentsRemoved = new List<EntityId>();
+
+private readonly AuthorityComparer authorityComparer = new AuthorityComparer();
+private readonly UpdateComparer<Update> updateComparer = new UpdateComparer<Update>();
+
+// Used to represent a state machine of authority changes. Valid state changes are:
+// authority lost -> authority lost temporarily
+// authority lost temporarily -> authority lost
+// authority gained -> authority gained
+// Creating the authority lost temporarily set is the aim as it signifies authority epoch changes
+private readonly HashSet<EntityId> authorityLost = new HashSet<EntityId>();
+private readonly HashSet<EntityId> authorityGained = new HashSet<EntityId>();
+private readonly HashSet<EntityId> authorityLostTemporary = new HashSet<EntityId>();
+
+private MessageList<ComponentUpdateReceived<Update>> updateStorage =
+    new MessageList<ComponentUpdateReceived<Update>>();
+
+private MessageList<AuthorityChangeReceived> authorityChanges =
+    new MessageList<AuthorityChangeReceived>();
+");
+
+                            foreach (var ev in eventDetailsList)
+                            {
+                                var eventType = $"{ev.EventName}.Event";
```

Comment:
I've seen this so often now, it should probably be a property on the `ev` type

Reply:
indeed - it should be :)